File: output_20260318_123118/gpt52/c/Linklist/map.c
#include "stdlib.h"
#include "assert.h"

struct node {
    struct node *next;
    int value;
};

/*@

predicate nodes(struct node *n; struct node *next, int value) =
    n != 0 &*& n->next |-> next &*& n->value |-> value;

predicate list(struct node *l; list<int> vs) =
    l == 0 &*& vs == nil
    ||
    l != 0 &*& nodes(l, ?n, ?v) &*& list(n, ?vs0) &*& vs == cons(v, vs0);

lemma void equals_reflexive(struct node *n)
    requires list(n, ?vs);
    ensures list(n, vs) &*& equals(n, n) == true;
{
    open list(n, vs);
    if (n != 0) {
        equals_reflexive(n->next);
    }
    close list(n, vs);
}

@*/

struct node *list_cons(int value, struct node *next)
    //@ requires list(next, ?vs);
    //@ ensures list(result, cons(value, vs));
{
    struct node *result = (struct node *)malloc(sizeof(struct node)); 
    if (result == 0) { abort(); }
    result->value = value;
    result->next = next;
    
    //@ close nodes(result, next, value);
    //@ close list(result, cons(value, vs));
    return result;
}

bool equals(struct node *n1, struct node *n2)
    //@ requires list(n1, ?vs1) &*& list(n2, ?vs2);
    //@ ensures list(n1, vs1) &*& list(n2, vs2) &*& result == (vs1 == vs2);
{
    
    
    bool result = false;
    if (n1 == 0)
        result = n2 == 0;
    else if (n2 == 0)
        result = false;
    else if (n1->value != n2->value)
        result = false;
    else {
        bool tmp = equals(n1->next, n2->next);
        result = tmp;
    }
    
    
    return result;
}

void dispose(struct node *l)
    //@ requires list(l, ?vs);
    //@ ensures true;
{
    
    if (l != 0) {
        //@ open list(l, vs);
        struct node *next = l->next;
        //@ open nodes(l, next, ?v);
        free(l);
        dispose(next);
    } else {
        //@ open list(l, vs);
    }
}

typedef int (* mapfunc)(void *data, int x);
    
    

struct node *fmap(struct node *list0, mapfunc f, void *data)
    //@ requires list(list0, ?vs);
    //@ ensures list(list0, vs) &*& list(result, ?ws);
{
    
    if (list0 == 0) {
        //@ open list(list0, vs);
        //@ close list(list0, nil);
        //@ close list(result, nil);
        return 0;
    } else {
        //@ open list(list0, vs);
        int fvalue = f(data, list0->value);
        struct node *fnext = fmap(list0->next, f, data);
        //@ close list(list0->next, ?vsTail);
        //@ close nodes(list0, list0->next, list0->value);
        //@ close list(list0, cons(list0->value, vsTail));
        
        struct node *result = list_cons(fvalue, fnext);
        
        return result;
    }
}

int plusOneFunc(void *data, int x) 
    //@ requires true;
    //@ ensures true;
{
    if (x == INT_MAX) abort();
    
    
    
    
    return x + 1;
}

int main() 
    //@ requires true;
    //@ ensures true;
{
    struct node *l = 0;
    //@ close list(l, nil);
    
    l = list_cons(3, l);
    l = list_cons(2, l);
    l = list_cons(1, l);
    
    struct node *l2 = fmap(l, plusOneFunc, 0);
    
    struct node *l3 = 0;
    //@ close list(l3, nil);
    
    l3 = list_cons(4, l3);
    l3 = list_cons(3, l3);
    l3 = list_cons(2, l3);
    bool tmp = equals(l2, l3);
    
    //@ assert tmp == ((cons(2, cons(3, cons(4, nil))) == cons(2, cons(3, cons(4, nil)))));
    assert(tmp);
    dispose(l);
    dispose(l2);
    dispose(l3);
    return 0;
}

File: output/gpt52/rust/purely_unsafe/solutions/map.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

type I32Func = unsafe fn(*mut u8, i32) -> i32;

/*@

predicate node(struct Node *n; struct Node *next, i32 value) =
    n->next |-> next &*& n->value |-> value;

predicate nodes(struct Node *n; list<i32> vs) =
    n == 0 ?
        vs == nil
    :
        node(n, ?next, ?v) &*& nodes(next, ?vs0) &*& vs == cons(v, vs0);

predicate stack(struct Stack *s; list<i32> vs) =
    s->head |-> ?h &*& nodes(h, vs);

predicate i32_cell(i32 *p; i32 v) =
    *p |-> v;

/* A functional contract for callbacks used by map_nodes/map. */
predicate i32_func(I32Func f; predicate(*mut u8;_) ctx) = true;

@*/

unsafe fn map_nodes(n: *mut Node, f: I32Func, data: *mut u8)
/*@ requires nodes(n, ?vs) &*& i32_func(f, ?ctx) &*& ctx(data, ?a0); @*/
/*@ ensures nodes(n, ?vs2) &*& i32_func(f, ctx) &*& ctx(data, ?a1); @*/
{
    if !n.is_null() {
        let y = f(data, (*n).value);
        (*n).value = y;
        map_nodes((*n).next, f, data);
    }
}

unsafe fn dispose_nodes(n: *mut Node)
/*@ requires nodes(n, ?vs); @*/
/*@ ensures true; @*/
{
    if !n.is_null() {
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    unsafe fn create() -> *mut Stack
    /*@ requires true; @*/
    /*@ ensures stack(result, nil); @*/
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();

        return stack;
    }

    unsafe fn push(stack: *mut Stack, value: i32)
    /*@ requires stack(stack, ?vs); @*/
    /*@ ensures stack(stack, cons(value, vs)); @*/
    {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
    }

    unsafe fn pop(stack: *mut Stack) -> i32
    /*@ requires stack(stack, cons(?v, ?vs0)); @*/
    /*@ ensures stack(stack, vs0) &*& result == v; @*/
    {
        let head = (*stack).head;

        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());

        return result;
    }

    unsafe fn map(stack: *mut Stack, f: I32Func, data: *mut u8)
    /*@ requires stack(stack, ?vs) &*& i32_func(f, ?ctx) &*& ctx(data, ?a0); @*/
    /*@ ensures stack(stack, ?vs2) &*& i32_func(f, ctx) &*& ctx(data, ?a1); @*/
    {
        map_nodes((*stack).head, f, data);
    }

    unsafe fn dispose(stack: *mut Stack)
    /*@ requires stack(stack, ?vs); @*/
    /*@ ensures true; @*/
    {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

unsafe fn plus_a(data: *mut u8, x: i32) -> i32
/*@ requires i32_cell(data as *mut i32, ?a); @*/
/*@ ensures i32_cell(data as *mut i32, a) &*& result == x + a; @*/
{
    let result = x + *(data as *mut i32);

    result
}

unsafe fn read_i32() -> i32
/*@ requires true; @*/
/*@ ensures true; @*/
{
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    line.parse().unwrap()
}

fn main()
/*@ requires true; @*/
/*@ ensures true; @*/
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        Stack::push(s, 30);
        let mut a = read_i32();
        //@ close i32_cell(&raw mut a, a);

        Stack::map(s, plus_a, &raw mut a as *mut u8);

        //@ open i32_cell(&raw mut a, _);
        Stack::dispose(s);
    }
}